### DISCUSSION TITLE
Using a 4G modem and GPS with Waydroid

### DISCUSSION THREAD

**Discussion by LukasGossmann** [waydroid/waydroid#1379] [Q&A]
So this might be a bit of an odd idea but i managed to get that partially working.
I recently bought an EC25 modem from Quectel with the idea to use it in conjunction with waydroid to basically have a complete android phone in a container.
While getting a RIL driver for this modem is a bit annoying (need to create account on Quectel forums and ask for support to send you a driver) i did manage to get it up and running with IRadio 1.1 capabilities pretty quickly.
That includes working 4G data, voice calls (currently with broken audio routing, if anyone knows more about audio routing in android please let me know), running USSD codes and even using the build in GPS receiver of this module.
So far so good however IRadio 1.1 is already pretty outdated and I was able to obtain an IRadio 1.6 capable driver.
This is where the issues start unfortunately:

First i encountered this linker error:
```CANNOT LINK EXECUTABLE "/vendor/bin/hw/rild": cannot locate symbol "_ZN7android8hardware12ProcessState16initWithMmapSizeEm"```

This turned out to be caused by the Waydroid patches adding an extra is `isHost` parameter to `ProcessState::initWithMmapSize` without ensuring the old method signature still exists.
That was a simple fix tough and got me a bit further:

```C++
sp<ProcessState> ProcessState::initWithMmapSize(size_t mmapSize) {
    return init(mmapSize, true /*requireMmapSize*/, false);
}
```
With that added the dirver was able to start properly and even started configuring the modem and registering itself with the HILD binder stuff (sorry i don't know what android calls all that).
That is where things go wrong though.....
It always crashes while performing a `BINDER_WRITE_READ ioctl`.
```
writev(3, [{iov_base="\1l\7T\257&f\177,\252$", iov_len=11}, {iov_base="\3", iov_len=1}, {iov_base="RILD\0", iov_len=5}, {iov_base="RIL_register_socket completed\0", iov_len=30}], 4) = 47
ioctl(16, BINDER_WRITE_READ, 0x7ffd5d7d9490) = 0
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0xffffffffffffffe0} ---
rt_sigaction(SIGSEGV, NULL, {sa_handler=0x7f944aac482a, sa_mask=~[KILL STOP], sa_flags=SA_RESTORER|SA_ONSTACK|SA_RESTART|SA_SIGINFO|0x800, sa_restorer=0x7f944ab71230}, 8) = 0
prctl(PR_GET_NO_NEW_PRIVS, 0, 0, 0, 0)  = 1
```
Since waydroid applies plenty of patches to everything binder related i was wondering if this crash might be caused by any of those:
(this is for LineageOS 20 with the patches from the folder base-patches-33)
```
0001-libcameraservice-Use-listByInterface.patch
0005-drm-Fix-drm-on-older-vendors.patch
0006-codec2-Fix-media-on-older-vendors.patch
0003-waydroid-core-jni-start-thread-pool-for-host-hwbinde.patch
0001-ServiceManagement-Use-host_hwbinder-if-available.patch
0002-Select-host_hwbinder-by-passing-argument-instead-of-.patch
0003-Don-t-wait-for-host-hwbinder.patch
0004-Always-re-new-ServiceManager.patch
0005-Fix-ABI-breakage-headers.patch
0007-Always-pass-useHostHwBinder-to-modified-methods-that.patch
0008-Add-support-for-host-hals-whitelist.patch
0001-hwbinder-Add-support-for-host_hwbinder-driver.patch
0001-waydroid-Allow-host_hwbinder-to-be-used-for-HALs.patch
```

The Quectel support is trying its best and even went ahead and downloaded the LineageOS source code and tried applying the waydroid patches but unfortunately failed because there were apparently to many conflicts at that moment.
I wasn't able to verify that since i had zero conflicts even after pulling the repos again but i can`t blame them for being a bit annoyed by all that already...

Anyways i was hoping that someone might know what exactly the host_hwbinder is used for as its not even a device file that i have inside the lxc container...
Would removing those patches cause any issues with waydroid ?
Also looking at it from the quectel support perspective: Would missing those patches cause a significant enough difference between their RIL driver and Waydroid when it comes to using the binder interface ? Apart from the missing symbol mentioned above it doesn't seem like anything else is conflicting (as in different method signatures)
And finally does anyone know of a way how i can better debug whats going on ?
Their RIL driver is closed source and i only get provided with some .so files to copy into /vendor/lib64

Any help / ideas are very much appreciated.
Thanks in advance :)

**Comment by fcodes0** [comment]
@LukasGossmann Any luck integrating Quecetel Modems within Waydroid? I'm trying to figure out how to build a custom image but have no clue how where the RIL files need to go. I ran the basic commands listed on waydroid docs to build a custom image based of lineage 18.1, but kind of lost on where to go from there. I found your thread on the Quecetel forums but after reading through it I ended up slightly confused with more questions. If you are still attempting this please let me know as I would love to connect further to try and get this running.

**Reply by LukasGossmann** [reply]
Hello @fcodes0,

i was partially successful in getting things to work. Partially meaning:
RIL 1.1 works with normal cellphone type stuff (sms, USSD codes, voice calls without sound) and cellular data, as well as GPS since my module also includes a GPS receiver.
Sound in calls does not work because of a limitation of the RIL driver. It only starts the call with AT commands but leaves it up to the end user on how to get sound into android. As far as i know the modem itself will output / input stuff on its dedicated audio pins though but i haven't tested that.

RIL 1.6 or greater doesn't really want to work because the quectel driver fails to sign in the modem into the network. I had a lengthy discussion with their support about that but gave up because it was moving forward way too slowly and i was getting annoyed by this constant back and forth.

All of this comes with a big downside though. At least in the LineageOS 20 version of the waydroid custom image they modified  the function `_ZN7android8hardware12ProcessState16initWithMmapSizeEm`  to have an extra boolean parameter which also renames it to `_ZN7android8hardware12ProcessState16initWithMmapSizeEmb`. This completely breaks compatibility with the quectel driver unless they specifically compile a build against the waydroid patched LineageOS source code.
For more information on that and on specifically who knows how to do that at quectel please see here:
https://forums.quectel.com/t/android-ril-and-gnss-driver-for-android-13-on-x86-64/31970

I also created a (private for now) repo and invited you to that. It contains the build files that you need to modify for making a build that contains the RIL driver as well as the configuration on the host side (LXC container) for the android to access the modem.

If you have any questions or need help with setting it up then let me know.

**Comment by LukasGossmann** [comment]
I just got around to making the aforementioned repository public.
In case anyone wants to try something similar this can be used as a reference:
https://github.com/LukasGossmann/WaydroidQuectelRIL
